### DISCUSSION TITLE
Tooling and popularity

### DISCUSSION THREAD

**Discussion by jemag** [grafana/tanka#812] [General]
This is a bit of a meta discussion, but I think important nonetheless. 

While Tanka helped quite a bit in using jsonnet with Kubernetes, I think there is still a significant lack in terms of tooling that prevents new users from easily adopting it.

I think 2 big items could really help ease users transition and favor adoption are:
**1. Proper completion and hover support by https://github.com/grafana/jsonnet-language-server
2. Integration with ArgoCD**

For point `1`, I really think this would be a game changer for new comers. Being able to complete unfamiliar code and see function signatures is something that can really ease the initial learning curve and is also something that is often expected from new users. There seems to be some preliminary work being done https://github.com/grafana/jsonnet-language-server/issues/5 however, most of the language server seems to be handled by a single developer. I am wondering if this is something that other people from within the Grafana organization might be interested in helping with, given the big benefits for jsonnet adoption?

For point `2`, I think this one is a bit more subtle, however I do think that ArgoCD is now one of the main (if not the most popular) GitOps tool in use at the moment. Being able to see Tanka as a first party integration alongside helm and kustomize, could be a welcoming sight and something that would also ease users into adopting it. While it is possible to do it with sidecar plugins, it is still a bit awkward and also is not the best experience if you have to deal with multiple private repositories. Seeing this issue https://github.com/argoproj/argo-cd/issues/3124, I was wondering if there is still interest from the Tanka community in providing ArgoCD first class support?

**Comment by Duologic** [comment]
The language server is great already, in neovim I have a shortcut that can peek into the definition, this is very useful to see what a function can do. It is just very hard to do with jsonnet given it is a lazy and untyped language.

For Argo, that is up for grabs. Within Grafana we run `tk export` and use Flux with the plain manifests in our clusters.

**Reply by jemag** [reply]
(deleted my previous comment as it contained couple mistakes). 

I use Neovim as well and I agree that `text/definition` does work quite well and overall julien did a great job with the language server, especially considering he developed most of it on his own. Although I still think completion and function signature would take it to the next level.

From what I can tell, I can indeed get the definition for example of deployment:
![image](https://user-images.githubusercontent.com/7985687/219147031-15133fab-3f34-495c-bc9c-c078c0717e58.png)
![image](https://user-images.githubusercontent.com/7985687/219147069-82e19736-52c0-4ec4-bb81-c79e300078fb.png)

However, you still have to know how to get there. For example (please let me know if you think this is actually already possible), I will get field completion after `k`:
![image](https://user-images.githubusercontent.com/7985687/219147473-1572ed30-8319-4e74-9fad-d00d6ac92a5a.png)
However no such completion afterwards
![image](https://user-images.githubusercontent.com/7985687/219147598-b1572b09-9b59-476d-b1bb-40199c3b5077.png)
![image](https://user-images.githubusercontent.com/7985687/219147667-27a76c04-670c-4068-9c27-315b4d4dc757.png)
So in this case I would need to type until `k.apps.v1.daemonSet` and then trigger `text/definition`. This would mean that the user needs to remember the full path or keep documentation open to get to `k.apps.v1.daemonSet`, instead of being able to rely on the completion.

It may not be that significant for experimented users, however I think these kind of small things ease the learning curve for new comers and may facilitate jsonnet's adoption.

**Reply by Duologic** [reply]
It's not perfect and lots of work to be done, feel free to help out if you like.

Here's my neovim LSPconfig, it might help: https://github.com/Duologic/nvim-config/blob/master/lua/configs/nvim-lspconfig.lua (notice `gp` mapping, this is what I was referring to)

**Reply by jemag** [reply]
Thanks I'll probably steal that `gp` mapping hehe. Any particular reason you don't use the `-t` flag with the jsonnet language server? This is what I have been using so far: https://github.com/jemag/dotfiles/blob/master/neovim/.config/nvim/lua/lsp/configs/jsonnnetls.lua

**Comment by Duologic** [comment]
Looks like I'm using a fork nvim-lspconfig: https://github.com/Duologic/nvim-lspconfig/blob/20a0a94f7fa22b19560ee50469f1f4c33b6922c8/lua/lspconfig/server_configurations/jsonnet_ls.lua
Seems like I set that up around the time we moved the language server to grafana org.
